server/src/main/java/com/thoughtworks/go/server/service/AuthorizationExtensionCacheService.java

```diff
@@ -0,0 +1,90 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+@Component
+public class AuthorizationExtensionCacheService {
+    private final int CACHE_EXPIRY_IN_MINUTES = 30;
+
+    private Cache<String, Boolean> isValidUserCache;
+    private Cache<String, AuthenticationResponse> getUserRolesCache;
+    private final AuthorizationExtension authorizationExtension;
+    private final String IS_VALID_USER_CACHE_KEY = "is_valid_user_request_for";
+
+    public AuthorizationExtensionCacheService(AuthorizationExtension authorizationExtension, Ticker ticker) {
+        this.authorizationExtension = authorizationExtension;
+        isValidUserCache = CacheBuilder.newBuilder()
+                .ticker(ticker).expireAfterWrite(CACHE_EXPIRY_IN_MINUTES, TimeUnit.MINUTES).build();
+        getUserRolesCache = CacheBuilder.newBuilder()
+                .ticker(ticker).expireAfterWrite(CACHE_EXPIRY_IN_MINUTES, TimeUnit.MINUTES).build();
+    }
+
+    @Autowired
+    public AuthorizationExtensionCacheService(AuthorizationExtension authorizationExtension) {
+        this(authorizationExtension, Ticker.systemTicker());
+    }
+
+    public boolean isValidUser(String pluginId, String username, SecurityAuthConfig authConfig) {
+        String cacheKey = cacheKeyFor(IS_VALID_USER_CACHE_KEY, pluginId, username, authConfig);
+        Boolean fromCache = isValidUserCache.getIfPresent(cacheKey);
+
+        if (fromCache == null) {
+            fromCache = authorizationExtension.isValidUser(pluginId, username, authConfig);
+            isValidUserCache.put(cacheKey, fromCache);
+        }
+
+        return fromCache;
+    }
+
+    public AuthenticationResponse getUserRoles(String pluginId, String username, List<SecurityAuthConfig> authConfigs, List<PluginRoleConfig> pluginRoleConfigs) {
+        String cacheKey = cacheKeyFor(IS_VALID_USER_CACHE_KEY, pluginId, username, authConfigs, pluginRoleConfigs);
+        AuthenticationResponse fromCache = getUserRolesCache.getIfPresent(cacheKey);
+
+        if (fromCache == null) {
+            fromCache = authorizationExtension.getUserRoles(pluginId, username, authConfigs, pluginRoleConfigs);
+            getUserRolesCache.put(cacheKey, fromCache);
+        }
+
+        return fromCache;
+    }
+
+    private String cacheKeyFor(String key, String pluginId, String username, List<SecurityAuthConfig> authConfigs, List<PluginRoleConfig> pluginRoleConfigs) {
+        String authConfigIds = authConfigs.stream().map(PluginProfile::getId).collect(Collectors.joining("__"));
+        String roleConfigNames = pluginRoleConfigs.stream().map(role -> role.getName().toString()).collect(Collectors.joining("__"));
+
+        return String.format("%s__%s__%s_auth_configs_%s_role_configs_%s", key, pluginId, username, authConfigIds, roleConfigNames);
```

Comment:
`__` are valid identifiers for all inputs that form a part of the cache key, which may allow an attacker some things to cause key conflicts to cause the cache to respond with something it should not.